---
title: "microbiome"
author: "Tao Wen(文涛)"
date: "`r Sys.Date()`"
output:
  html_document:
    df_print: paged
    theme: cerulean
    highlight: haddock
    toc: yes
    toc_depth: 3
    toc_float:
      collapsed: no
      smooth_scroll: yes
    code_fold: show
  word_document:
    toc: yes
    toc_depth: '3'
  pdf_document:
    toc: yes
    toc_depth: '3'
editor_options: 
  chunk_output_type: console
---


```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = T, echo=T, comment="#>", message=F, warning=F,
	fig.align="center", fig.width=7, fig.height=5, dpi=150)
```



```{css include=FALSE}
 pre code,pre,code {
 white-space:pre!important;
 overflow-x: scroll!important; 
} 
```



## Introducrion



The microbiome R package facilitates exploration and analysis of microbiome profiling data, in particular 16S taxonomic profiling.

This vignette provides a brief overview with example data sets from published microbiome profiling studies (Lahti et al. 2014, Lahti et al. (2013), O’Keefe et al. (2015)). A more comprehensive tutorial is available on-line.



### 导入数据

```{R}


library(microbiome)
library(phyloseq)
library(gcookbook)
library(tidyverse)
library(knitr)
library(ggClusterNet)
data(ps)

theme_set(theme_bw(21))

```


### 微生物组数据基本统计描述

统计测序数据的数据量的最大最小值，均值，中位数；OTU中书否包含singleton及其比例。

```{R}

summarize_phyloseq(ps)


```


### alpha多样性

microbiome极大的扩展了alpha多样性的内容

这里首次将alpha分的这么细，随着近年来网络的风靡，和关键微生物类群的文章层出不穷，多样性增加了许多指标。
丰富性，均匀性，多样性，支配性（dominance）和稀有性（rarity）的各种度量。有关参数调整的更多选项，请参见各个功能
注意：这里均是默认的参数运行的全部alpha多样性指标
一共有22中指标;
这里有两种指标包含在phyloseq中，使用 estimate_richness函数计算，所以这里不能计算这两个指标。

```{r}
tab <- microbiome::alpha(ps, index = "all")
kable(head(tab))
#-单独计算richness
tab <- richness(ps)
kable(head(tab))
#-单独计算支配物种指数
tab <- dominance(ps, index = "all")
kable(head(tab))
#-计算稀有微生物指数
tab <- rarity(ps, index = "all")
kable(head(tab))
#-计算覆盖度指标
tab <- coverage(ps, threshold = 0.5)
kable(head(tab))
#-计算核心微生物指数
tab <- core_abundance(ps, detection = .1/100, prevalence = 50/100)
kable(head(tab))
#-计算基尼系数
tab <- inequality(ps)
kable(head(tab))
#-计算均匀度指数
tab <- evenness(ps, "all")
kable(head(tab))
#--计算alpha多样性指数差异显著性
d <- meta(ps)
d$diversity <- diversities(ps, "shannon")$diversity_shannon
# 按照分组分开指数
spl <- split(d$diversity, d$Group)
# Kolmogorov-Smironv test
pv <- ks.test(spl$KO, spl$OE)$p.value
# 矫正p-value
padj <- p.adjust(pv)

```

### beta多样性排序


### 微生物组样本分布分析

这部分虽然官方文档显示支持phyloseq的全部排序方式，但是显然在使用的过程中PCA ，t-SNE，CAP不太好用。但是phyloseq支持的距离这里都是能用的。
- https://microbiome.github.io/tutorials/Landscaping.html


```{R}

# Convert to compositional data
pseq <- microbiome::transform(ps, "compositional")

# Pick core taxa
pseq.core <- core(pseq, detection = 1/1000, prevalence = 5/100)
# #--PCA ，t-SNE，CAP不可用
# plot_landscape(pseq, method = "PCA",  col = "Group") +
#        labs(title = paste("PCA / CLR"))

p <- plot_landscape(pseq,
                      method = "NMDS", distance = "bray", col = "Group") +
       labs(title = paste("NMDS / Compositional / Bray-Curtis"))
p
#-使用需要进化树计算的距离
p1 <- plot_landscape(microbiome::transform(pseq.core, "compositional"),
                      method = "PCoA", distance = "unifrac", col = "Group") +
       labs(title = paste("PCoA / Compositional / unifrac")) + 
  theme_bw()
print(p1)
path = "./2_Microbiome data analysis integrated class R package/pipeline/02.microbiome/"
ggsave(paste(path,"Fig1_PCoA.pdf",sep = ""),p1,width = 6,height = 5)
#---如何提取数据，手动作图

set.seed(423542)
x <- pseq.core
quiet(x.ord <- ordinate(x, "NMDS", "bray"))
# Pick the projected data (first two columns + metadata)
proj <- phyloseq::plot_ordination(x, x.ord, justDF=TRUE)
# Rename the projection axes
names(proj)[1:2] <- paste("Comp", 1:2, sep=".")

# 使用Microbioem包出图
p <- plot_landscape(proj[, 1:2], col = proj$nationality, legend = T)
print(p)
head(proj)
#-提取数据自己出图
p <- ggplot(aes_string(x = proj$Comp.1, y = proj$Comp.2, label = "Group"), data = proj) +
       geom_text(size = 2)
print(p)

```


### 门水平微生物群落丰度展示

```{R}
ps1 <- ps %>%
  # subset_samples(Group == "OE") %>%
  tax_glom_wt(ranks = "Phylum") %>%  
  microbiome::transform(transform = "compositional")

p2 <- ps1 %>%
  plot_composition(sample.sort = "Bacteroidetes", otu.sort = "abundance") 
print(p2)

ggsave(paste(path,"./Fig2_abundance_bar.pdf",sep = ""),p2,width = 10,height = 8)

```


### 微生物组成展示-核心微生物提取和可视化

提取相对丰度超过0.01的微生物，核心微生物提取和可视化：https://microbiome.github.io/tutorials/Core.html；

core函数用于提取核心微生物群落数据，detection代表提取相对丰度大于某个阈值的微生物。prevalence 代表出现额概率超过百分之多少的样本。

```{R}

pseq =  microbiome::transform(ps, "compositional") %>% 
  core(detection = 0.01, prevalence = 50/100)

# With compositional (relative) abundances
det <- c(0, 0.1, 0.5, 2, 5, 20)/100
prevalences <- seq(.05, 1, .05)

p <- pseq %>%
  plot_composition(sample.sort = "Firmicutes", otu.sort = "abundance") 

print(p)



library(RColorBrewer)

prevalences <- seq(.05, 1, .05)
detections <- 10^seq(log10(1e-3), log10(.2), length = 10)

# Also define gray color palette
gray <- gray(seq(0,1,length=5))
#pseq.rel<- microbiome::transform(pseq, 'compositional')
p <- plot_core(pseq, plot.type = "heatmap", colours = gray,
       prevalences = prevalences, detections = detections) +
    labs(x = "Detection Threshold (Relative Abundance (%))")
p
ggsave(paste(path,"./Fig2_core_heat.pdf",sep = ""),p,width = 10,height = 6)



detections <- 10^seq(log10(1), log10(max(abundances(pseq))/10), length = 10)

p <- plot_core(pseq, plot.type = "heatmap", 
             prevalences = prevalences,
             detections = detections,
         colours = rev(brewer.pal(5, "Spectral")),
         min.prevalence = .2, horizontal = TRUE) 
print(p)

ggsave(paste(path,"./Fig2_OTU_heat.pdf",sep = ""),p,width = 14,height = 10)




```

### 物种共有和特有情况分析可视化

物种存在与否做一个Ven图：
https://microbiome.github.io/tutorials/core_venn.html

```{R}
library(eulerr)
library(microbiome)

pseq.rel <- microbiome::transform(ps, "compositional")
disease_states <- unique(as.character(meta(pseq.rel)$Group))
print(disease_states)

list_core <- c() # an empty object to store information

for (n in disease_states){ # for each variable n in DiseaseState
    #print(paste0("Identifying Core Taxa for ", n))
    
    ps.sub <- subset_samples(pseq.rel, Group == n) # Choose sample from DiseaseState by n
    
    core_m <- core_members(ps.sub, # ps.sub is phyloseq selected with only samples from g 
                           detection = 0.001, # 0.001 in atleast 90% samples 
                           prevalence = 0.75)
    print(paste0("No. of core taxa in ", n, " : ", length(core_m))) # print core taxa identified in each DiseaseState.
    list_core[[n]] <- core_m # add to a list core taxa for each group.
    #print(list_core)
}

mycols <- c(nonCRC="#d6e2e9", CRC="#cbf3f0", H="#fcf5c7") 

# ggsave(paste(path,"Fig2_OTU_heat.pdf",sep = ""),p,width = 14,height = 10)


pdf(paste(path,"/Fig3_ven.pdf",sep = ""))
plot(venn(list_core),
     fills = mycols)

dev.off()

```

### 使用热图展示微生物丰度信息

https://microbiome.github.io/tutorials/Heatmap.html

```{R}
library(reshape2)
# Pick data subset (DI samples from Phylum Bacteroidetes)
# tax_table(ps)[,1:6]
pseq2 <- ps %>%
         subset_taxa(Genus == "Ramlibacter")

# Z transformed abundance data
pseqz <- microbiome::transform(pseq2, "Z")

# Plot the abundances heatmap
dfm <- melt(abundances(pseqz))
colnames(dfm) <- c("Taxa", "Sample", "value")
# ggsave(paste(path,"Fig2_OTU_heat.pdf",sep = ""),p,width = 14,height = 10)
pdf("./Fig4_heatmap.pdf",width = 15,height = 12)
heat(dfm, "Taxa", "Sample", "value")
dev.off()


# Sort the matrix rows and cols directly
xo <- neat(abundances(pseqz), method = "NMDS", distance = "euclidean") 

# Heatmap visualization
dfm <- melt(xo)
colnames(dfm) <- c("Taxa", "Sample", "value")
heat(dfm, "Taxa", "Sample", "value")
sorted.rows <- neatsort(abundances(pseqz), "rows", method = "NMDS", distance = "euclidean") 
```


### 指定微生物丰富展示和差异分析

https://microbiome.github.io/tutorials/post_hoc.html

```{R}
p4 <- boxplot_abundance(ps, x = "Group", y = "ASV_1") + scale_y_log10()
print(p4)
# ggsave(paste(path,"Fig2_OTU_heat.pdf",sep = ""),p,width = 14,height = 10)
ggsave(paste(path,"./Fig5_special_microbe.pdf",sep = ""),p4,width = 4,height = 4)

library(tidyr)

# Analyse specific taxa
tax <- "ASV_1"

# Pick the signal (abundance) for this tax
sample_data(ps)$signal <- get_sample(ps, tax)


attach(meta(ps))
res<- pairwise.wilcox.test(signal, Group, p.adjust.method = 'BH')
resK<- kruskal.test(signal, Group)
print(resK)



```


### 使用DEsep2做差异分析

此为外界端口

https://microbiome.github.io/tutorials/deseq2.html

```{R}

library(magrittr)


# Only check the core taxa to speed up examples
pseq <- core(ps, detection = 50, prevalence = 80/100)

library(phyloseq)
library(reshape2)
library(DESeq2)
library(knitr)

# Running the DESeq2 analysis
ds2 <- phyloseq_to_deseq2(pseq, ~ Group)
dds <- DESeq(ds2)
res <- results(dds)
df <- as.data.frame(res)
df$taxon <- rownames(df)
df <- df %>% arrange(log2FoldChange, padj)


print(head(kable((df))))

# Identify top taxa based on standard ANOVA
source(system.file("extdata/check_anova.R", package = "microbiome"))
ano <- check_anova(pseq, "Group");
ano$log2FC <- log2(ano$ave.KO) - log2(ano$ave.OE)
taxa.anova <- as.character(subset(ano, padj < 0.01 & abs(log2FC) > log2(2))$taxa)

# lowPick the top taxa based on DESEq2
taxa.deseq <- subset(res, padj < 0.01 & abs(log2FoldChange) > log2(2))$taxon

# Check overlap
# Most DESEq2 taxa are confirmed with ANOVA


# Also the est p-values are well correlated (higher not so)
mf <- data.frame(df$padj, ano$padj)
p <- ggplot(mf, aes(x = log10(df$padj), y = log10(ano$padj))) +
  labs(x = 'DESeq2 adjusted p-value', y = 'ANOVA adjusted p-value') +
  geom_point()
print(p)

```

### limma 和t检验

外界端口

https://microbiome.github.io/tutorials/limma.html

```{R}
# Load libraries
library(microbiome)
library(ggplot2)
library(dplyr)


map = sample_data(ps)
map$ID = row.names(map)
sample_data(ps) = map

pssub =  subset_samples(ps,!Group %in% c("OE"));pssub 

otu <- abundances(microbiome::transform(pssub, "log10"))
meta <- meta(pssub)

# Compare the two groups with limma
library(limma)

# Prepare the design matrix which states the groups for each sample
# in the otu
design <- cbind(intercept = 1, Grp2vs1 = meta[["Group"]])
rownames(design) <- rownames(meta)
design <- design[colnames(otu), ]

head(design)
# NOTE: results and p-values are given for all groupings in the design matrix
# Now focus on the second grouping ie. pairwise comparison
coef.index <- 2

# Fit the limma model
fit <- lmFit(otu, design)
fit <- eBayes(fit)

# Limma P-values
pvalues.limma = fit$p.value[, 2]

# Limma effect sizes
efs.limma <-  fit$coefficients[, "Grp2vs1"]

# Summarise
library(knitr)
kable(topTable(fit, coef = coef.index, p.value=0.1), digits = 2)

# QQ
qqt(fit$t[, coef.index], df = fit$df.residual + fit$df.prior); abline(0,1)

# Volcano
volcanoplot(fit, coef = coef.index, highlight = coef.index)

# Compare the two groups with t-test
library(dplyr)
pvalues.ttest <- c()
male.samples <- dplyr::filter(meta, Group == "WT")$ID
female.samples <- dplyr::filter(meta, Group == "KO")$ID
for (tax in rownames(otu)) {
  pvalues.ttest[[tax]] <- t.test(otu[tax, male.samples], otu[tax, female.samples])$p.value
}
# Multiple testing correction
pvalues.ttest <- p.adjust(pvalues.ttest, method = "fdr")

# Compare p-values between limma and t-test
taxa <- rownames(otu)
plot(pvalues.ttest[taxa], pvalues.limma[taxa])
abline(0,1,lty = 2)


```


### 群落差异检测

外接端口

```{R}
# Pick relative abundances (compositional) and sample metadata 
pseq.rel <- microbiome::transform(ps, "compositional")
otu <- abundances(pseq.rel)
meta <- meta(pseq.rel)

p <- plot_landscape(pseq.rel, method = "NMDS", distance = "bray", col = "Group", size = 3)
print(p)
# samples x species as input
library(vegan)
permanova <- adonis(t(otu) ~ Group,
                    data = meta, permutations=99, method = "bray")

# P-value
print(as.data.frame(permanova$aov.tab)["Group", "Pr(>F)"])


```



```{R}

sessionInfo()


```


