---
title: "生成各级物种分类汇总——STAMP输入文件"
author: "刘永鑫(Yong-Xin Liu)"
date: '`r Sys.Date()`'
output:
  html_document:
    df_print: paged
    theme: cerulean
    highlight: haddock
    toc: yes
    toc_depth: 3
    toc_float:
      collapsed: no
      smooth_scroll: yes
    code_fold: show
  word_document:
    toc: yes
    toc_depth: '3'
  pdf_document:
    toc: yes
    toc_depth: '3'
--- 

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
 
## 读取输入文件

```{r read}
# 1. 读取OTU表
otutab = read.table("otutab.txt", header=T, row.names= 1, sep="\t", comment.char = "", stringsAsFactors = F)
# 2. 读取物种注释
tax = read.table("taxonomy.txt", header=T, row.names= 1, sep="\t",comment.char = "", stringsAsFactors = F) 
# OTU丰度筛选阈值，默认0.01%，0为来筛选
thre = 0.01
# 输出文件名前缀
prefix = "stamp/tax"

# 生成各分类级汇总特征表
suppressWarnings(suppressMessages(library(amplicon)))
format2stamp(otutab, tax, thre, prefix)
# 在当前目录生成tax_1-8共7个级别+OTU过滤文件
```
使用此脚本，请引用下文：

If used this script, please cited:

**Yong-Xin Liu**, Yuan Qin, **Tong Chen**, et. al. A practical guide to amplicon and metagenomic analysis of microbiome data. **Protein Cell**, 2021(12) 5:315-330, doi: [10.1007/s13238-020-00724-8](https://doi.org/10.1007/s13238-020-00724-8)

Copyright 2016-2022 Yong-Xin Liu <metagenome@126.com>