---
title: "MicrobiomeAnalystR"
author: "Tao Wen(文涛)"
date: "`r Sys.Date()`"
output:
  html_document:
    df_print: paged
    theme: cerulean
    highlight: haddock
    toc: yes
    toc_depth: 3
    toc_float:
      collapsed: no
      smooth_scroll: yes
    code_fold: show
  word_document:
    toc: yes
    toc_depth: '3'
  pdf_document:
    toc: yes
    toc_depth: '3'
editor_options: 
  chunk_output_type: console
---


```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = T, echo=T, comment="#>", message=F, warning=F,
	fig.align="center", fig.width=7, fig.height=5, dpi=150)
```



### R包安装

MicrobiomeAnalystR安装需要依赖非常多的R包，这也是其功能强大的原因之一。并且很多的R包都在github，有些长期以来维护较少，并且中国用户往往难以下载安装，
这些困难让我们安装MicrobiomeAnalystR不是很容易。


```{R eval=FALSE, include=FALSE}

#pacman 管理R包，瑞国不存在可以直接安装
library(pacman)

pacman::p_load(phyloseq, metacoder, pryr, biomformat, RColorBrewer, ggplot2, gplots, Cairo, igraph,
               BiocParallel, randomForest, metagenomeSeq, MASS, DESeq2, vegan, RJSONIO, ggfortify, pheatmap, xtable, genefilter,
               data.table, reshape, stringr, ape, grid, gridExtra, splitstackshape, edgeR, globaltest, R.utils, viridis, ggrepel,
               ppcor)
# BiocManager::install("qs")
# BiocManager::install('metacoder')
# BiocManager::install('pryr')
# BiocManager::install('metagenomeSeq')
# BiocManager::install('globaltest')
# devtools::install_github("cran/ppcor")
# devtools::install_github("kylebittinger/qiimer")
# devtools::install_github("kylebittinger/qiimer")
# install.packages("./Tax4Fun-master/", repos = NULL, type = "source")
# devtools::install_github("xia-lab/MicrobiomeAnalystR", build = TRUE, build_opts = c("--no-resave-data", "--no-manual"))

```



### 数据导入

```{R}

library("MicrobiomeAnalystR")

path = "./2_Microbiome data analysis integrated class R package/pipeline/03.MicrobiomeAnalystR/"

#-初始化项目
mbSet<-Init.mbSetObj()
mbSet<-SetModuleType(mbSet, "mdp")#--扩增子数据标记为："mdp"


mbSet<-ReadSampleTable(mbSet, "./2_Microbiome data analysis integrated class R package//data_MicrobiomeAnalystR/map.txt");

mbSet<-Read16STaxaTable(mbSet, "../../data_MicrobiomeAnalystR/taxonomy.txt");
mbSet<-Read16SAbundData(mbSet, "../../data_MicrobiomeAnalystR/otutab.txt","text","Others/Not_specific","T");
mbSet <- ReadTreeFile(mbSet,"../../data_MicrobiomeAnalystR/otus.tree")


```

### 基本统计

样本序列数量统计可视化

```{R}
mbSet<-PlotLibSizeView(mbSet, "norm_libsizes_0","pdf");
names(mbSet)
```

### 挑选样本和微生物数据过滤

ApplyAbundanceFilter函数过滤OTU表格，下面函数保留至少read数超过4，并且至少出现在百分之20的OTU。
ApplyVarianceFilter：根据变异系数过滤掉在样本之间变化不大的OTU。

```{R}
#---转化为phyloseq对象进行分析
mbSet<-CreatePhyloseqObj(mbSet, "text","Others/Not_specific","F")
mbSet<-SanityCheckData(mbSet, "text");

mbSet<-ApplyAbundanceFilter(mbSet, "prevalence", 4, 0.2);
mbSet<-ApplyVarianceFilter(mbSet, "iqr", 0.1);
# 过滤样本，选择需要分析的样本进行后续统计分析
smpl.nm.vec <- c("KO1","KO2","KO3","KO4","KO5","KO6","OE1","OE2","OE3","OE4","OE5","OE6","WT1","WT2","WT3","WT4","WT5","WT6")
mbSet<-UpdateSampleItems(mbSet);


```

### 微生物组数据标准化

PerformNormalization提供三个层面的数据标准化：

- 数据抽平
- 数据标准化，这里转录组常用的矫正测序数据的标准化方法都支持：TSS，CSS，UQ等。
- 数据转化：提供了RLE，TMM，CLR三种方法，这里是调用了EDger包中的三种方法。

```{R}
# ?PerformNormalization
mbSet<-PerformNormalization(mbSet, "rarewi", "colsum", "rle");
```


### 群落物种组成分析展示

PlotTaxaAundanceBar函数提供七个分类等级的堆叠柱状图和面积图展示微生物组组成信息，指定分组变量"SampleType"，默认合并序列数量小于10的微生物来减少颜色分配，设定输出格式为png。结果会导出图片和数据表格。

```{R}
# ?PlotTaxaAundanceBar
# 抽平
mbSet<-PlotTaxaAundanceBar(mbSet,"taxa_alpha_0","Phylum","SampleType","null","barraw",10,"set3","sum",10, "bottom", "F",format ="pdf");
# 修改为相对丰度堆叠柱状图
mbSet<-PlotTaxaAundanceBar(mbSet, "taxa_alpha_1","Class","SampleType", "none", "barnorm",10, "set3","sum",10, "bottom", "F", "pdf");
# 转换堆叠柱状图为横向
mbSet<-PlotTaxaAbundanceBarSamGrp(mbSet, "taxa_alpha_2","Class","SampleType", "none", "barnorm",10,"set3","med", 10, "bottom", "F", "pdf");
#-通过设定toptaxa大小来调整多少个低丰度的微生物被合并
mbSet<-PlotTaxaAbundanceBarSamGrp(mbSet, "taxa_alpha_3","Phylum","SampleType", "none", "barnorm",1000,"set3","med", 10, "bottom", "F", "pdf")

```

### 物种组成-饼图分析

网页上，这里的是交互式饼图。这里我们用于保存。

```{R}
mbSet<-PlotOverallPieGraph(mbSet, "Phylum", 10,"sum", 10, "bottom");
GetSeriesColors()
mbSet<-SavePiechartImg(mbSet, "Phylum","primary_piechart_0","pdf");

#-通过设定toptaxa大小来调整多少个低丰度的微生物被合并
mbSet<-PlotOverallPieGraph(mbSet, "Phylum", 100,"sum", 10, "bottom");
GetSeriesColors()
mbSet<-SavePiechartImg(mbSet, "Phylum","primary_piechart_1","pdf");
```

### alpha多样性分析

绘制稀释曲线

```{R}
mbSet<-PlotRarefactionCurve(mbSet, "filt","SampleType","SampleType","SampleType","5","rarefaction_curve_0","pdf");

```

alpha多样性指标计算和可视化：
PlotAlphaData：提供单个样本alpha多样性分析和点线图展示；
PlotAlphaBoxData：提供分组alpha展示；
PerformAlphaDiversityComp：提供alpha多样性统计检验。


```{R}
# mbSet<-PlotPhylogeneticTree(mbSet, "SampleType","SampleType","Phylum","rectangular","phylogenetic_tree_0","png");

mbSet<-PlotAlphaData(mbSet, "filt","alpha_diver_0","Chao1","SampleType","OTU", "default", "pdf");
mbSet<-PlotAlphaBoxData(mbSet, "alpha_diverbox_0","Chao1","SampleType","default", "pdf");
mbSet<-PerformAlphaDiversityComp(mbSet, "tt","SampleType");

```


### 进化树可视化

```{R}
#
mbSet<-PlotPhylogeneticTree(mbSet, "SampleType","SampleType","Phylum","rectangular","phylogenetic_tree_0","pdf");
#-环状
mbSet<-PlotPhylogeneticTree(mbSet, "SampleType","SampleType","Phylum","radial","phylogenetic_tree_1","pdf");

```

### beta多样性排序和群落差异检测

PerformBetaDiversity函数已经不支持了，换成了PlotBetaDiversity。使用方法一样。

```{R}

mbSet<-PlotBetaDiversity(mbSet, "beta_diver_0","PCoA","bray","expfac","SampleType","none","OTU","","Chao1", "yes", "pdf", 72, "default");#
mbSet<-PCoA3D.Anal(mbSet, "PCoA","bray","OTU","expfac","SampleType","","Chao1","beta_diver3d_0.json")
mbSet<-PerformCategoryComp(mbSet, "OTU", "adonis","bray","SampleType");
#-样本聚类
mbSet<-PlotTreeGraph(mbSet, "plot_tree_0","bray","ward.D","SampleType","OTU", "default", "pdf");


#更换群落差异检测方法
mbSet<-PerformCategoryComp(mbSet, "OTU", "permdisp","bray","SampleType");
mbSet$analSet$stat.info
```



### 微生物组差异分析


#### metagenomeSeq差异

```{R}


# 差异分析metagenomeSeq
mbSet<-ApplyAbundanceFilter(mbSet, "prevalence", 4, 0.2);
mbSet<-ApplyVarianceFilter(mbSet, "iqr", 0.1);
mbSet<-PerformNormalization(mbSet, "none", "colsum", "none");
mbSet<-PerformMetagenomeSeqAnal(mbSet, "SampleType",0.05,"NA","OTU","zigfit");

```


#### lefse


```{R}

mbSet<-PerformLefseAnal(mbSet,  0.1, "fdr", 2.0,  "SampleType","F","NA","OTU");
mbSet<-PlotLEfSeSummary(mbSet, 15, "dot",  "bar_graph_0","png");# 图片没有内容
mbSet<-PerformUnivarTest(mbSet, "SampleType",0.05,"NA","OTU","tt");

```



###  微生物丰度可视化

```{R}
# # 核心微生物分析
# mbSet<-CoreMicrobeAnalysis(mbSet, "core_micro_0",0.2,0.01,"OTU","bwm","overview", "all_samples", "SampleType", "null", "pdf");

# 热图展示微生物丰度
mbSet<-PlotHeatmap(mbSet, "heatmap_0","euclidean","ward.D","bwm","SampleType","OTU","overview","F", "pdf","T","F");
# 从门水平上展示丰度
mbSet<-PlotHeatmap(mbSet, "heatmap_1","euclidean","ward.D","bwm","SampleType","Phylum","overview","F", "pdf","T","F");

```

### 微生物组样本聚类分析

```{R}
#-聚类分析，指定按照分组上色
mbSet<-PlotTreeGraph(mbSet, "plot_tree_0","bray","ward.D","SampleType","OTU", "default", "pdf");

```


###微生物组机器学习

```{R}
mbSet<-RF.Anal(mbSet, 500,7,1,"SampleType","OTU")
mbSet<-PlotRF.Classify(mbSet, 15, "rf_cls_0","pdf", width=NA)
mbSet<-PlotRF.VIP(mbSet, 15, "rf_imp_0","pdf", width=NA)

```



### 附-有些函数无法跑出来-有问题


```{R}

# mbSet<-PrepareCorrExpValues(mbSet, "SampleType", "Phylum", "dbgr", "reingold-tilford", "all", "0.05")# 没有这个函数
# mbSet<-PerformNetworkCorrelation(mbSet,"Phylum", "sparcc", "expr",100, 0.05, 0.3, "mean", "cor_net_0.json")# 报错
# mbSet<-PerformRNAseqDE(mbSet, "EdgeR",0.05,"SampleType","NA","OTU");# 报错
# RNA方法运行差异分析不可运行
# mbSet<-PerformRNAseqDE(mbSet, "EdgeR",0.05,"SampleType","NA","OTU")
```


```{R}
sessionInfo()
```
